File: Yikang's/Price Track/SPY_VXX.py
import matplotlib as plt
import os
import pandas as pd
import numpy as np
from tqdm import *
from datetime import timedelta
import math

def to_nanosecond(time):
    time_array = time.split(":")
    return ((int(time_array[0]) * 3600 + int(time_array[1]) * 60 + float(time_array[2])) * 1000000)

def move_na(object_array):
    for i in range(len(object_array)):
        if math.isnan(object_array[i]) is False:
            last = object_array[i]
        else:
            object_array[i] = last
    return(object_array)


os.chdir("D:/SkyDrive/Documents/UIUC/CME Fall 2016")
SPY = pd.read_table('data/price/SPY_price_track.txt', sep='', delimiter='\t')
VXX = pd.read_table('data/price/VXX_price_track.txt', sep='', delimiter='\t')
SPY['Time'] = SPY['Time'].map(lambda x: to_nanosecond(x))
VXX['Time'] = VXX['Time'].map(lambda x: to_nanosecond(x))


SPY.drop(['Unnamed: 0', 'Quantity'], axis=1, inplace=True)
VXX.drop(['Unnamed: 0', 'Quantity'], axis=1, inplace=True)
SPY.columns = ['Time', 'SPY Price']
VXX.columns = ['Time', 'VXX Price']
SPY['SPY Net'] = SPY['SPY Price']/SPY['SPY Price'][0]
VXX['VXX Net'] = VXX['VXX Price']/VXX['VXX Price'][0]


merged = pd.merge(SPY, VXX, how='outer')
merged[['Time', 'SPY Net', 'VXX Net']].plot()
merged.sort_values('Time', inplace=True)
merged.reset_index(inplace=True)
merged['SPY Net'][0] = merged['SPY Net'][4]
merged['VXX Net'] = move_na(merged.copy()['VXX Net'])
merged['SPY Net'] = move_na(merged.copy()['SPY Net'])
merged.set_index('Time', drop = True, inplace = True)
merged[['SPY Net', 'VXX Net']].plot()

File: Yikang's/Price Track/SPY_VXX.R
require(quantmod)
require(ggplot2)
options(digits = 12)
setwd("/Users/luoy2/OneDrive/Documents/UIUC/CME Fall 2016")
SPY <- read.table(paste(getwd(), "/data/price/SPY_price_track.txt", sep = ''), 
                  sep = '\t',header = TRUE, nrow = FALSE, stringsAsFactors = FALSE)
VXX <- read.table(paste(getwd(), "/data/price/VXX_price_track.txt", sep = ''), 
                  sep = '\t',header = TRUE, nrow = FALSE, stringsAsFactors = FALSE)

time.convert <- function(x){
        group <- strsplit(x, ":")[[1]]
        h <- group[1]
        m <- group[2]
        s <- group[3]
        as.numeric(h)*3600 + as.numeric(m)*60 + as.numeric(s)
}
return.get <- function(x){
        ROC(x, type = 'discrete')
}

SPY$Return <- return.get(SPY$Price)
VXX$Return <- return.get(VXX$Price)
SPY$Time <- sapply(SPY$Time, time.convert)
VXX$Time <- sapply(VXX$Time, time.convert)
merged <- merge(SPY[c('Time', 'Return')], VXX[c('Time', 'Return')],
                by = 'Time', 
                all = TRUE,
                suffixes = c('.SPY', '.VXX'))
merged[is.na(merged)] <- 0

correlation_test <- function(lag){
        time <- merged[,1]
        vec1 <- merged[,2]
        vec2 <- c(rep(NA, lag), merged[,3])[1:nrow(merged)]
        time_gap <- time[lag+1] - time[1]
        c(time_gap, cor(vec1, vec2, use = "complete"))
}

result <- data.frame(t(mapply(correlation_test, 1:10000)))
result[which(result$X2 == min(result$X2)),]

ggplot(data = result, aes(x=X1, y=X2))+ labs(x = 'Lag(s)', y = 'Correlation') + geom_line()
